### PR TITLE
fabtests/common: add option to configure message sizes 

### PR DESCRIPTION
**Pull request by mmubarak**
The change adds an option to specify a range of messages
to fabtests by using the format 'r:start,increment,end'
to the "-S" option. Prior to this change, we were able to
provide "all" or a specific size to the message size option.

Updates have been made to pytest/efa tests to use this option
for message sizes in the range of 8k-9k with an increment of 4.

Tests were conducted using runfabtests.py to ensure the message
ranges are being processed by pytest (cuda testing is not done).

### CODE REVIEW

File: fabtests/common/shared.c [hunk 1]
```diff
@@ -119,7 +119,7 @@ struct fi_rma_iov remote;
 
 struct ft_opts opts;
 
-struct test_size_param test_size[] = {
+struct test_size_param pre_test_size[] = {
```
Comment (wzamazon) [inline fabtests/common/shared.c:122]: it is not clear what `pre` means. maybe just name the variable `all_test_size` because it is used by `-S all`?

File: fabtests/common/shared.c [hunk 2]
```diff
@@ -3076,6 +3114,9 @@ void ft_parsecsopts(int op, char *optarg, struct ft_opts *opts)
 	case 'S':
 		if (!strncasecmp("all", optarg, 3)) {
 			opts->sizes_enabled = FT_ENABLE_ALL;
+		} else if (!strncasecmp("r:", optarg, 2)){
+			opts->sizes_enabled = FT_ENABLE_ALL;
```
Comment (wzamazon) [inline fabtests/common/shared.c:3116]: We might need to rename `FT_ENABLE_ALL` to something else.

File: fabtests/pytest/efa/test_rdm.py [hunk 3]
```diff
@@ -24,18 +24,36 @@ def run_rdm_test(cmdline_args, executable, iteration_type, completion_type, memo
                           pytest.param("standard", marks=pytest.mark.standard)])
 def test_rdm_pingpong(cmdline_args, iteration_type, completion_type, memory_type):
     run_rdm_test(cmdline_args, "fi_rdm_pingpong", iteration_type,
-                 completion_type, memory_type)
+            completion_type, memory_type, "all")
+
+@pytest.mark.parametrize("message_size",
+                         [pytest.param("r:8000,4,9000", marks=pytest.mark.short)])
```
Comment (wzamazon) [inline fabtests/pytest/efa/test_rdm.py:29]: I do not think we need the "short" marker here. With the "short" marker, the test will no be run when user specify "-t standard".

**Review by wzamazon — COMMENTED** [review]
Overall looks good to me. Let some minor comments.

Also:

The commit title should have a ":" to separate "fatests/common" and "add ....".

Also the first commit is unnecessary, probably need a rebase.

**Comment by mmubarak** [comment]
> 



> Can the python scripts just call the existing tests in a loop changing the size? I thought that was one of the reasons to add python scripts.

For this specific case, we are doing a parameter sweep of a large range of message sizes. The performance overhead with python scripts running the test with 250 message sizes is 4362 secs (1500 tests executed) whereas if we directly update the fabtests, it runs 16 tests and the runtime is only 106 secs. That's why for this case, we prefer to update fabtests.

File: fabtests/pytest/efa/test_rdm.py (hunk 3 above)
Comment (mmubarak) [reply fabtests/pytest/efa/test_rdm.py:29]: Isn't this the idea to run these message sizes with short marker only? We don't want to enable them with long iterations.

File: fabtests/common/shared.c [hunk 4]
```diff
@@ -3061,6 +3063,42 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+       /* use string tokenizer */
+       const int MAX_TOKS = 3;
+       int inputs [MAX_TOKS];
+       const char* delim = ",";
+       char* token = strtok (rem_str, delim);
+       int j = 0;
+
+       while (token != NULL)
+       {
+	 inputs[j] = atoi (token);
```
[the remark this answers is not in the document]
Comment (mmubarak) [reply fabtests/common/shared.c]: Fixed

File: fabtests/common/shared.c (hunk 2 above)
Comment (mmubarak) [reply fabtests/common/shared.c]: Renamed to FT_SIZES_ALL

File: fabtests/common/shared.c [hunk 5]
```diff
@@ -1598,6 +1602,9 @@ void ft_free_res(void)
 
 	ft_close_fids();
 
+	if (range_test_size) {
+		free (range_test_size);
+	}
```
Comment (shefty) [inline fabtests/common/shared.c:1607]: free() accepts a null pointer, no need to check.  Please check coding style.  The open paren should be next to free - no space.  Comment applies to other locations in the code.

File: fabtests/common/shared.c [hunk 6]
```diff
@@ -3059,6 +3066,44 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+	       assert (j < MAX_TOKS);
+	       /* parse next token */
+	       token = strtok (NULL, delim);
+	       j++;
+       }
+       int start = inputs[0], inc = inputs[1], end = inputs[2];
+
+       /* populate the array */
+       int i;
+       int test_range_cnt = (end - start) / inc + 1;
```
Comment (wzamazon) [inline fabtests/common/shared.c:3095]: maybe name it "range_test_cnt" to be consistent with "range_test_size"

File: fabtests/common/shared.c [hunk 7]
```diff
@@ -2922,7 +2929,7 @@ void ft_csusage(char *name, char *desc)
 	FT_PRINT_OPTS_USAGE("-I <number>", "number of iterations");
 	FT_PRINT_OPTS_USAGE("-Q", "bind EQ to domain (vs. endpoint)");
 	FT_PRINT_OPTS_USAGE("-w <number>", "number of warmup iterations");
-	FT_PRINT_OPTS_USAGE("-S <size>", "specific transfer size or 'all'");
+	FT_PRINT_OPTS_USAGE("-S <size>", "specific transfer size or a range of sizes (syntax r:start,inc,end) or 'all'");
```
Comment (shefty) [inline fabtests/common/shared.c:2932]: line length should be limited to 80 characters.  Comment applies to other places in the code.

This would be simpler as a couple new options, rather than creating a complex parsing string.  -E end_size -T size stepping.  If -E is omitted, end_size = start size (-S option).  If -T omitted, stepping is by 1

File: fabtests/common/shared.c [hunk 8]
```diff
@@ -3059,6 +3066,44 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+
+       /* use string tokenizer */
+       const int MAX_TOKS = 3;
+       int inputs [MAX_TOKS];
+       const char* delim = ",";
+       char* token = strtok (rem_str, delim);
+       int j = 0;
+
+       while (token != NULL)
+       {
```
Comment (shefty) [inline fabtests/common/shared.c:3084]: open braces should be at end of previous line

File: fabtests/common/shared.c [hunk 9]
```diff
@@ -3059,6 +3066,44 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+	       inputs[j] = atoi (token);
+	       assert (j < MAX_TOKS);
+	       /* parse next token */
+	       token = strtok (NULL, delim);
+	       j++;
+       }
+       int start = inputs[0], inc = inputs[1], end = inputs[2];
```
Comment (shefty) [inline fabtests/common/shared.c:3091]: place all variable definitions at the top of the function.  Mingling definitions throughout a function makes them hard to find, making maintenance more difficult.

File: fabtests/include/shared.h [hunk 10]
```diff
@@ -78,11 +78,11 @@ struct test_size_param {
 	int enable_flags;
 };
 
-extern struct test_size_param test_size[];
-extern const unsigned int test_cnt;
+extern struct test_size_param * test_size;
```
Comment (shefty) [inline fabtests/include/shared.h:81]: * should be next to variable name

File: fabtests/include/shared.h [hunk 11]
```diff
@@ -78,11 +78,11 @@ struct test_size_param {
...
+extern unsigned int test_cnt;
 #define TEST_CNT test_cnt
 
-#define FT_ENABLE_ALL		(~0)
+#define FT_ENABLE_SIZES		(~0)
```
Comment (shefty) [inline fabtests/include/shared.h:85]: please separate renaming existing items into its own commit.

File: fabtests/common/shared.c (hunk 5 above)
Comment (mmubarak) [reply fabtests/common/shared.c]: Done

File: fabtests/common/shared.c (hunk 7 above)
Comment (mmubarak) [reply fabtests/common/shared.c]: I have updated the changed lines in this PR to be less than 80 characters-- however, I haven't updated rest of the lines in the code that were violating the 80 character constraint (I saw a bunch while using the marker).

File: fabtests/common/shared.c (hunk 8 above)
Comment (mmubarak) [reply fabtests/common/shared.c]: Done

File: fabtests/common/shared.c (hunk 9 above)
Comment (mmubarak) [reply fabtests/common/shared.c]: Agreed, done.

File: fabtests/include/shared.h (hunk 11 above)
Comment (mmubarak) [reply fabtests/include/shared.h:85]: Done.

File: fabtests/common/shared.c (hunk 7 above)
Comment (mmubarak) [reply fabtests/common/shared.c]: Regarding using -E and -T for end size and stepping, it would be simpler indeed but there are two caveats (1) The code currently sets FT_OPT_SIZE flag when -S is used with a transfer size. The benchmarks make specific checks if FT_OPT_SIZE is not enabled then a range or all sizes are used. This approach would require modifications to the flag values and the benchmarks if we want to use -S as a range. (2) We are already using -E for out of band address exchange and we didn't want to run out of characters to be used.

Comment (shefty) [reply fabtests/common/shared.c]: I picked random characters for the options.  We would obviously need to pick ones that work.  Other issues can easily be solved.  If an end size is set, just clear FT_OPT_SIZE flag, or add a new flag to indicate a specific range is requested instead, or use a different option to specify the starting value.

**Comment by wzamazon** [comment]
The title of the commit that changed the pytest test case need some tweak. Please add efa to it to mark it as an EFA specific change. like "fabtests/pytest/efa: xxxx".

File: fabtests/common/shared.c (hunk 7 above)
Comment (wzamazon) [reply fabtests/common/shared.c]: There are not many letters left though.

I did some grep and noticed that fabtests have already used 34,  which means there are only 18 left. Maybe we want to save some for the future?


```
	FT_PRINT_OPTS_USAGE("-B <src_port>", "non default source port number");
	FT_PRINT_OPTS_USAGE("-P <dst_port>", "non default destination port number");
	FT_PRINT_OPTS_USAGE("-s <address>", "source address");
	FT_PRINT_OPTS_USAGE("-b[=<oob_port>]", "enable out-of-band address exchange and "
	FT_PRINT_OPTS_USAGE("-E[=<oob_port>]", "enable out-of-band address exchange only "
	FT_PRINT_OPTS_USAGE("-C <number>", "number of connections to accept before "
	FT_PRINT_OPTS_USAGE("-F <addr_format>", "Address format (default:FI_FORMAT_UNSPEC)");
	FT_PRINT_OPTS_USAGE("-f <fabric>", "fabric name");
	FT_PRINT_OPTS_USAGE("-d <domain>", "domain name");
	FT_PRINT_OPTS_USAGE("-p <provider>", "specific provider name eg sockets, verbs");
	FT_PRINT_OPTS_USAGE("-e <ep_type>", "Endpoint type: msg|rdm|dgram (default:rdm)");
	FT_PRINT_OPTS_USAGE("-U", "run fabtests with FI_DELIVERY_COMPLETE set");
	FT_PRINT_OPTS_USAGE("-M <mode>", "Disable mode bit from test");
	FT_PRINT_OPTS_USAGE("-K", "fork a child process after initializing endpoint");
	FT_PRINT_OPTS_USAGE("-a <address vector name>", "name of address vector");
	FT_PRINT_OPTS_USAGE("-h", "display this help output");
	FT_PRINT_OPTS_USAGE("-f <fabric>", "fabric name");
	FT_PRINT_OPTS_USAGE("-d <domain>", "domain name");
	FT_PRINT_OPTS_USAGE("-p <provider>", "specific provider name eg sockets, verbs");
	FT_PRINT_OPTS_USAGE("-D <device_iface>", "Specify device interface: "
	FT_PRINT_OPTS_USAGE("-i <device_id>", "Specify which device to use (default: 0)");
	FT_PRINT_OPTS_USAGE("-H", "Enable provider FI_HMEM support");
	FT_PRINT_OPTS_USAGE("-h", "display this help output");
	FT_PRINT_OPTS_USAGE("-I <number>", "number of iterations");
	FT_PRINT_OPTS_USAGE("-Q", "bind EQ to domain (vs. endpoint)");
	FT_PRINT_OPTS_USAGE("-w <number>", "number of warmup iterations");
	FT_PRINT_OPTS_USAGE("-S <size>", "specific transfer size or 'all'");
	FT_PRINT_OPTS_USAGE("-l", "align transmit and receive buffers to page size");
	FT_PRINT_OPTS_USAGE("-m", "machine readable output");
	FT_PRINT_OPTS_USAGE("-D <device_iface>", "Specify device interface: "
	FT_PRINT_OPTS_USAGE("-i <device_index>", "Index of the device to use");
	FT_PRINT_OPTS_USAGE("-t <type>", "completion type [queue, counter]");
	FT_PRINT_OPTS_USAGE("-c <method>", "completion method [spin, sread, fd, yield]");
	FT_PRINT_OPTS_USAGE("-h", "display this help output");
```

File: fabtests/pytest/efa/test_rdm.py [hunk 12]
```diff
@@ -23,18 +23,36 @@ def run_rdm_test(cmdline_args, executable, iteration_type, completion_type, memo
                           pytest.param("standard", marks=pytest.mark.standard)])
 def test_rdm_pingpong(cmdline_args, iteration_type, completion_type, memory_type):
     run_rdm_test(cmdline_args, "fi_rdm_pingpong", iteration_type,
-                 completion_type, memory_type)
+            completion_type, memory_type, "all")
+
+@pytest.mark.parametrize("message_size",
+                         [pytest.param("r:8000,4,9000", marks=pytest.mark.functional)])
```
[the remark this answers is not in the document]
Comment (mmubarak) [reply fabtests/pytest/efa/test_rdm.py:29]: The motivation to keep this parameterized is that we are able to see the range as parameter while the tests are running. If we use a simple marker, we can't distinguish between tests that are taking range as arguments.

[the remark this answers is not in the document]
Comment (wzamazon) [reply fabtests/pytest/efa/test_rdm.py:29]: I see, make sense

**Comment by shijin-aws** [comment]
The sign-off the commits should be `FirstName LastName <email>`. Could you also use amazon email?

File: fabtests/common/shared.c (hunk 7 above)
Comment (a-szegel) [reply fabtests/common/shared.c]: I think we should create a new argument called 

```
--msg_size=(500,700,4]
```

and implement custom parser logic (or use an existing library) to translate the provided CLI range string into python range object 

```
msg_size = range(504,704,4)
```

This way the proposed `--msg_size` argument will be extensible to all fabtests, and offer more clarity on its purpose.

Comment (mmubarak) [reply fabtests/common/shared.c]: We do want to have these tests written in C for performance reasons. Having more flexibility to parse fabtest arguments by using string instead of int might be useful in the long run.

For now I have used sscanf to parse the range using the -S option. We thought about introducing a different '-R' option to parse this range but it would be counter-intuitive to provide together "-S all -R [start, inc, end]" for message sizes.

File: fabtests/pytest/efa/test_rdm.py [hunk 13]
```diff
@@ -23,18 +23,36 @@ def run_rdm_test(cmdline_args, executable, iteration_type, completion_type, memo
...
 @pytest.mark.parametrize("iteration_type",
                          [pytest.param("short", marks=pytest.mark.short),
                           pytest.param("standard", marks=pytest.mark.standard)])
 def test_rdm_tagged_pingpong(cmdline_args, iteration_type, completion_type, memory_type):
     run_rdm_test(cmdline_args, "fi_rdm_tagged_pingpong", iteration_type,
-                 completion_type, memory_type)
+                 completion_type, memory_type, "all")
+
+@pytest.mark.parametrize("message_size",
+                         [pytest.param("r:8000,4,9000", marks=pytest.mark.functional)])
```
Comment (wzamazon) [inline fabtests/pytest/efa/test_rdm.py:42]: Can add 2 more ranges:  [0,4,64] (where we cross device's inline range), and [4048,4,4148] (around 4096, which is shm protocol cross point)

Comment (mmubarak) [reply fabtests/pytest/efa/test_rdm.py]: Done

File: fabtests/pytest/efa/test_rdm.py [hunk 14]
```diff
@@ -23,18 +23,42 @@ def run_rdm_test(cmdline_args, executable, iteration_type, completion_type, memo
                           pytest.param("standard", marks=pytest.mark.standard)])
 def test_rdm_pingpong(cmdline_args, iteration_type, completion_type, memory_type):
     run_rdm_test(cmdline_args, "fi_rdm_pingpong", iteration_type,
-                 completion_type, memory_type)
+            completion_type, memory_type, "all")
+
+@pytest.mark.parametrize("message_size",
+                         [pytest.param("r:8000,4,9000", marks=pytest.mark.functional),
+                          pytest.param("r:0,4,64", marks=pytest.mark.functional),
+                          pytest.param("r:4048,4,4148", marks=pytest.mark.functional)])
```
Comment (wzamazon) [inline fabtests/pytest/efa/test_rdm.py:31]: Since this range are used repeatedly, we can define a fixture for it.

See line 9 `memory_type` for example.

An minor issue is that when you try to use the new fixture in `test_rma_bw`, you will need to explicitly import it.

So I suggest to create a new file `efa_common.py` and put the new fixture and `memory_type` there.

File: fabtests/common/shared.c [hunk 15]
```diff
@@ -119,7 +119,7 @@ struct fi_rma_iov remote;
 
 struct ft_opts opts;
 
-struct test_size_param test_size[] = {
+struct test_size_param all_test_size[] = {
```
Comment (shefty) [inline fabtests/common/shared.c:122]: In the future, it would be better to separate out name changes like this as a separate patch.  Don't worry about that change for this.

File: fabtests/common/shared.c [hunk 16]
```diff
@@ -146,7 +146,11 @@ struct test_size_param test_size[] = {
 	{ 1 << 23, 0 },
 };
 
-const unsigned int test_cnt = (sizeof test_size / sizeof test_size[0]);
+unsigned int test_cnt = (sizeof all_test_size / sizeof all_test_size[0]);
+
+struct test_size_param * test_size = all_test_size;
+/* range of messages is dynamically allocated */
+struct test_size_param * range_test_size = NULL;
```
Comment (shefty) [inline fabtests/common/shared.c:153]: nit: Above definitions should have the '*' next to the variable name - no space in between.  Yes, when you're reading code for hours on end each day, little things like this matter.

global vars will be initialized to 0/NULL, so no need to set.  Harmless though if you want to keep it.

File: fabtests/common/shared.c [hunk 17]
```diff
@@ -1597,7 +1601,7 @@ void ft_free_res(void)
 	rx_ctx_arr = NULL;
 
 	ft_close_fids();
-
+	free (range_test_size);
```
Comment (shefty) [inline fabtests/common/shared.c:1604]: nit: remove space between free and '('

File: fabtests/common/shared.c [hunk 18]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
 	}
 }
 
+void parse_opts_range (char* optarg)
+{
+       int start, inc, end;
```
Comment (shefty) [inline fabtests/common/shared.c:3069]: start, end, and inc should be size_t

File: fabtests/common/shared.c [hunk 19]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
 	}
 }
 
+void parse_opts_range (char* optarg)
```
Comment (shefty) [inline fabtests/common/shared.c:3067]: remove space after function name

File: fabtests/common/shared.c [hunk 20]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+{
+       int start, inc, end;
+       int i, range_test_cnt;
```
Comment (shefty) [inline fabtests/common/shared.c:3070]: is range_test_cnt needed?  Can't we just set test_cnt?

File: fabtests/common/shared.c [hunk 21]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+       sscanf (optarg, "r:%d,%d,%d", &start, &inc, &end);
+
+       /* populate the array */
```
Comment (shefty) [inline fabtests/common/shared.c:3073]: drop comment - it's obvious from the code that's what is happening.

File: fabtests/common/shared.c [hunk 22]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
 	}
 }
 
+void parse_opts_range (char* optarg)
+{
+       int start, inc, end;
+       int i, range_test_cnt;
+       sscanf (optarg, "r:%d,%d,%d", &start, &inc, &end);
```
Comment (shefty) [inline fabtests/common/shared.c:3071]: remove space after function name.  Add check for return value to validate that input was correct and return a failure code if not.

File: fabtests/common/shared.c [hunk 23]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+
+       /* populate the array */
+       range_test_cnt = (end - start) / inc + 1;
```
Comment (shefty) [inline fabtests/common/shared.c:3074]: This calculation is off if end doesn't align with start + some multiple of inc.  E.g. start = 1, end = 2, and inc = 2 has 1 valid value.  There needs to be some adjustment to end so that it falls on a valid value.  Btw, I like the idea of this supporting decreasing sizes, but that's added complexity that may not be worth adding right now.  If not, then we'll want a sanity check that end >= start

File: fabtests/common/shared.c [hunk 24]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+       TEST_CNT = range_test_cnt;
+       range_test_size = malloc (sizeof (struct test_size_param) *
```
Comment (shefty) [inline fabtests/common/shared.c:3076]: remove space after function name.  it would be safer here to use calloc() and avoid the initialization to 0 below.  As a general rule, sizeof() should be paired with a variable name, not a structure definition.  E.g.
```
range_test_size = calloc(test_cnt, sizeof(*range_test_size));
```
Add a check for allocation failure

File: fabtests/common/shared.c [hunk 25]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+       int i, range_test_cnt;
+       sscanf (optarg, "r:%d,%d,%d", &start, &inc, &end);
+
+       /* populate the array */
+       range_test_cnt = (end - start) / inc + 1;
+       TEST_CNT = range_test_cnt;
+       range_test_size = malloc (sizeof (struct test_size_param) * 
+		       range_test_cnt);
+       for (i = 0; i < range_test_cnt && i < end; i++) {
+	       size_t val = start + (i * inc);
```
Comment (shefty) [inline fabtests/common/shared.c:3079]: val isn't needed

File: fabtests/common/shared.c [hunk 26]
```diff
@@ -3059,6 +3064,25 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+	       range_test_size [i].size = val;
+	       range_test_size [i].enable_flags = 0;
```
Comment (shefty) [inline fabtests/common/shared.c:3081]: remove space after array name before [

File: fabtests/include/shared.h (hunk 10 above)
Comment (shefty) [reply fabtests/include/shared.h:81]: github changed the comment when submitting to make it a bulleted list.  The '*' should be next to the variable name -- *test_name.

File: fabtests/include/shared.h [hunk 27]
```diff
@@ -78,8 +78,8 @@ struct test_size_param {
 	int enable_flags;
 };
 
-extern struct test_size_param test_size[];
-extern const unsigned int test_cnt;
+extern struct test_size_param * test_size;
+extern unsigned int test_cnt;
 #define TEST_CNT test_cnt
```
Comment (shefty) [inline fabtests/include/shared.h:83]: We should revisit why we have TEST_CNT instead of just using test_cnt --

File: fabtests/include/shared.h [hunk 28]
```diff
@@ -508,7 +508,7 @@ void show_perf(char *name, size_t tsize, int iters, struct timespec *start,
 		struct timespec *end, int xfers_per_iter);
 void show_perf_mr(size_t tsize, int iters, struct timespec *start,
 		struct timespec *end, int xfers_per_iter, int argc, char *argv[]);
-
+void parse_opts_range (char *optarg);
```
Comment (shefty) [inline fabtests/include/shared.h:511]: remove space after function name

**Review by shefty — CHANGES_REQUESTED** [review]
See comments on first patch.  Thanks!  General comment -- save typing - avoid the space bar. :)

File: fabtests/pytest/efa/test_rdm.py (hunk 14 above)
Comment (mmubarak) [reply fabtests/pytest/efa/test_rdm.py:31]: Parameterize is essentially doing the same thing here. I don’t see the value in moving to fixtures, it adds further complexity IMO.

Comment (wzamazon) [reply fabtests/pytest/efa/test_rdm.py:31]: It is the similar, but have it as fixture do remove duplicated code, and we might want to add more ranges in the future.

File: fabtests/include/shared.h (hunk 10 above)
Comment (mmubarak) [reply fabtests/include/shared.h]: > See comments on first patch. Thanks! General comment -- save typing - avoid the space bar. :)

I have made the changes requested. However, I do think it is quite some manual work both for reviewer and developer to inspect code style changes. Many C/C++ based libraries offer python scripts that serve two purposes (1) mark the CI as failed if there is a code styling violation and (2) re-format the developer code into appropriate code style. The check_style.py script used by NS-3 is a nice example that does both https://www.nsnam.org/doxygen/group___check_style.html

File: fabtests/pytest/efa/test_rdm.py (hunk 14 above)
Comment (mmubarak) [reply fabtests/pytest/efa/test_rdm.py]: Shared fixtures have to be declared in conftest.py and I moved them to conftest as recommended in pytest documentation. However, our conftest.py is global to pytest and in the long run, we should think about having plug-ins to import fixtures from individual modules.

File: fabtests/pytest/conftest.py [hunk 29]
```diff
@@ -210,3 +210,13 @@ def prefix_type(request):
...
 def datacheck_type(request):
     return request.param
+
+@pytest.fixture(scope="module", params=["host_to_host", "host_to_cuda",
+                                        "cuda_to_host", "cuda_to_cuda"])
+def memory_type(request):
+    return request.param
+
+@pytest.fixture(scope="module", params=["r:8000,4,9000", 
+                                        "r:0,4,64", "r:4048,4,4148"])
```
[the remark this answers is not in the document]
Comment (mmubarak) [reply fabtests/pytest/conftest.py:220]: The pytest recommendation is to have all shared fixtures defined in a single conftest.py file https://gist.github.com/peterhurford/09f7dcda0ab04b95c026c60fa49c2a68

**Review by wzamazon — APPROVED** [review]
Thank you!

**Comment by mmubarak** [comment]
@shefty do you have any more comments on this?

File: fabtests/common/shared.c (hunk 15 above)
Comment (shefty) [reply fabtests/common/shared.c:122]: Actually, let's rename this to def_test_sizes rather than all.

File: fabtests/common/shared.c [hunk 30]
```diff
@@ -146,7 +146,11 @@ struct test_size_param test_size[] = {
 	{ 1 << 23, 0 },
 };
 
-const unsigned int test_cnt = (sizeof test_size / sizeof test_size[0]);
+unsigned int test_cnt = (sizeof all_test_size / sizeof all_test_size[0]);
+
+struct test_size_param* test_size = all_test_size;
+/* range of messages is dynamically allocated */
+struct test_size_param* range_test_size;
```
Comment (shefty) [inline fabtests/common/shared.c:153]: The '*' needs to go next to the variable name, not the structure name.

File: fabtests/common/shared.c [hunk 31]
```diff
@@ -3059,6 +3064,29 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
 	}
 }
 
+void parse_opts_range(char* optarg)
+{
+       size_t start, inc, end;
+       int i;
+       int ret = sscanf(optarg, "r:%ld,%ld,%ld", &start, &inc, &end);
```
Comment (shefty) [inline fabtests/common/shared.c:3071]: Please don't mix functionality with variable declarations, and add a space after the declarations.
```
int i, ret;

ret = sscanf(...);
if (ret != 3) ...
```

File: fabtests/common/shared.c [hunk 32]
```diff
@@ -3059,6 +3064,29 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
...
+       if (ret != 3) {
+	      perror("sscanf");
+	      exit (EXIT_FAILURE);
```
Comment (shefty) [inline fabtests/common/shared.c:3074]: extra space

File: fabtests/common/shared.c [hunk 33]
```diff
@@ -3059,6 +3064,29 @@ void ft_parse_hmem_opts(int op, char *optarg, struct ft_opts *opts)
 	}
 }
 
+void parse_opts_range(char* optarg)
```
Comment (shefty) [inline fabtests/common/shared.c:3067]: add ft_ prefix to function name

File: fabtests/include/shared.h [hunk 34]
```diff
@@ -78,8 +78,8 @@ struct test_size_param {
 	int enable_flags;
 };
 
-extern struct test_size_param test_size[];
-extern const unsigned int test_cnt;
+extern struct test_size_param* test_size;
```
Comment (shefty) [inline fabtests/include/shared.h:81]: wrong placement for *

**Comment by mmubarak** [comment]
@shefty do you have any more feedback for this?
